apps/air-discount-scheme/backend/src/app/modules/flight/flight.service.ts

```diff
@@ -62,6 +66,186 @@ export class FlightService {
+    // from the second intended chronological leg
+    if (
+      REYKJAVIK_FLIGHT_CODES.includes(secondFlight.origin) ||
+      REYKJAVIK_FLIGHT_CODES.includes(firstFlight.destination)
+    ) {
+      delta = -delta
+    }
+
+    if (delta >= 0 && delta <= CONNECTING_FLIGHT_GRACE_PERIOD) {
+      return true
+    }
+    return false
+  }
+
+  async isFlightLegConnectingFlight(
+    existingFlightId: string,
+    incomingLeg: FlightLeg,
+  ): Promise<boolean> {
+    // Get the corresponding flight for the connection discount code
+    const existingFlight = await this.flightModel.findOne({
+      where: {
+        id: existingFlightId,
+      },
+      include: [
+        {
+          model: this.flightLegModel,
+          where: { financialState: availableFinancialStates },
+        },
+      ],
+    })
+
+    if (!existingFlight) {
+      return false
+    }
+
+    // If a user flightLeg exists such that the incoming flightLeg makes a valid connection
+    // pair, return true
+    for (const flightLeg of existingFlight.flightLegs) {
+      if (
+        this.hasConnectingFlightPotentialFromFlightLegs(flightLeg, incomingLeg)
+      ) {
+        return true
+      }
+    }
+
+    return false
+  }
+
+  async countThisYearsConnectedFlightsByNationalId(
+    nationalId: string,
+  ): Promise<number> {
+    const currentYear = new Date(Date.now()).getFullYear().toString()
+
+    const noConnectedFlightLegs = await this.flightModel.count({
+      where: Sequelize.and(
+        Sequelize.where(
+          Sequelize.fn(
+            'date_part',
+            'year',
+            Sequelize.fn('date', Sequelize.col('booking_date')),
+          ),
+          currentYear,
+        ),
+        { nationalId },
+      ),
+      include: [
+        {
+          model: this.flightLegModel,
+          where: {
+            financialState: availableFinancialStates,
+            isConnectingFlight: true,
+          },
+        },
+      ],
+    })
+
+    const noConnectableFlightLegs = await this.flightModel.count({
+      where: Sequelize.and(
+        Sequelize.where(
+          Sequelize.fn(
+            'date_part',
+            'year',
+            Sequelize.fn('date', Sequelize.col('booking_date')),
+          ),
+          currentYear,
+        ),
+        { nationalId },
+      ),
+      include: [
+        {
+          model: this.flightLegModel,
+          where: {
+            financialState: availableFinancialStates,
+            [Op.or]: [
+              { origin: { [Op.in]: REYKJAVIK_FLIGHT_CODES } },
+              { destination: { [Op.in]: REYKJAVIK_FLIGHT_CODES } },
+            ],
+          },
+        },
+      ],
+    })
+
+    return noConnectableFlightLegs - noConnectedFlightLegs
+  }
+
+  async findThisYearsConnectableFlightsByNationalId(
+    nationalId: string,
+  ): Promise<Flight[]> {
+    const currentYear = new Date(Date.now()).getFullYear().toString()
+
+    let flights = await this.flightModel.findAll({
```

Comment:
To make the code more concise we could use `findThisYearsFlightsByNationalId` and just use a filter for the `connectable` part.

Reply:
Good point :+1: